File: android/app/src/main/java/com/example/yugiohscanner/ml/PriceSteps.kt
package com.example.yugiohscanner.ml

data class StepPoint(val day: String, val price: Double)
data class StepMarker(val day: String, val family: String)
data class Steps(
    val kind: String,
    val flatDay: String? = null,
    val flatPrice: Double? = null,
    val points: List<StepPoint> = emptyList(),
    val markers: List<StepMarker> = emptyList(),
)

/**
 * Spec G1 §4.5 -- Stufenlinie des Preisverlaufs.
 * ZWILLING: desktop/src/utils/priceSteps.js. Beide laufen gegen docs/fixtures/portfolio/price-steps.json.
 * Wer eine Seite aendert, aendert beide.
 */
object PriceSteps {
    fun compute(rows: List<PriceRef>, today: String, windowDays: Int): Steps {
        val sorted = rows.sortedBy { it.day }
        if (sorted.isEmpty()) return Steps("none")
        if (sorted.size == 1) return Steps("flat", flatDay = sorted[0].day, flatPrice = sorted[0].price)
        val start = UtcDay.add(today, -windowDays)
        val points = ArrayList<StepPoint>()
        val markers = ArrayList<StepMarker>()
        var before: PriceRef? = null
        for (i in sorted.indices) {
            val r = sorted[i]
            if (r.day < start) { before = r; continue }
            if (r.day > today) continue
            val b = before
            if (points.isEmpty() && b != null && r.day != start) points += StepPoint(start, b.price)
            points += StepPoint(r.day, r.price)
            if (i > 0) {
                val prevFam = PriceFamily.ofSource(sorted[i - 1].source)
                val fam = PriceFamily.ofSource(r.source)
                if (prevFam != fam) markers += StepMarker(r.day, fam)
            }
        }
        val b = before
        if (points.isEmpty() && b != null) points += StepPoint(start, b.price)
        val last = points.lastOrNull()
        if (last != null && last.day != today) points += StepPoint(today, last.price)
        return Steps("series", points = points, markers = markers)
    }
}

File: desktop/src/utils/priceSteps.js
import FAMILIES from '../../electron/price-families.json' with { type: 'json' };

// Spec G1 §4.5 — Stufenlinie des Preisverlaufs.
// ZWILLING: android/app/src/main/java/com/example/yugiohscanner/ml/PriceSteps.kt. Beide laufen gegen
// docs/fixtures/portfolio/price-steps.json. Wer eine Seite aendert, aendert beide.

export const familyOfSource = (source) => FAMILIES[source] || 'unknown';
export const FAMILY_LABELS = { cm: 'Cardmarket', ygo: 'YGOPRODeck', manual: 'manuell', unknown: 'unbekannt' };

export const todayUtc = (now = new Date()) => now.toISOString().slice(0, 10);

export function addDaysUtc(day, n) {
  const d = new Date(`${day}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + n);
  return d.toISOString().slice(0, 10);
}

export const fmtDayDE = (day) => `${day.slice(8, 10)}.${day.slice(5, 7)}.`;

export function computeSteps(rows, today, windowDays) {
  const sorted = [...(rows || [])].sort((a, b) => (a.day < b.day ? -1 : a.day > b.day ? 1 : 0));
  if (sorted.length === 0) return { kind: 'none', points: [], markers: [] };
  if (sorted.length === 1) {
    return { kind: 'flat', flatDay: sorted[0].day, flatPrice: Number(sorted[0].price), points: [], markers: [] };
  }
  const start = addDaysUtc(today, -windowDays);
  const points = [];
  const markers = [];
  let before = null;
  for (let i = 0; i < sorted.length; i++) {
    const r = sorted[i];
    if (r.day < start) { before = r; continue; }
    if (r.day > today) continue;
    if (points.length === 0 && before && r.day !== start) points.push({ day: start, price: Number(before.price) });
    points.push({ day: r.day, price: Number(r.price) });
    const prev = sorted[i - 1];
    if (prev && familyOfSource(prev.source) !== familyOfSource(r.source)) {
      markers.push({ day: r.day, family: familyOfSource(r.source) });
    }
  }
  if (points.length === 0 && before) points.push({ day: start, price: Number(before.price) });
  const last = points[points.length - 1];
  if (last && last.day !== today) points.push({ day: today, price: last.price });
  return { kind: 'series', points, markers };
}
